File: izinga-ordermanager/src/main/java/io/curiousoft/izinga/ordermanagement/service/ProfileServiceImpl.java
package io.curiousoft.izinga.ordermanagement.service;

import io.curiousoft.izinga.commons.model.Profile;
import io.curiousoft.izinga.commons.model.ProfileRoles;
import io.curiousoft.izinga.commons.profile.events.ProfileCreatedEvent;
import io.curiousoft.izinga.commons.profile.events.ProfileDeletedEvent;
import io.curiousoft.izinga.commons.profile.events.ProfileUpdatedEvent;
import io.curiousoft.izinga.commons.repo.ProfileRepo;
import org.springframework.beans.BeanUtils;
import org.springframework.context.ApplicationEventPublisher;

import jakarta.validation.ConstraintViolation;
import jakarta.validation.Validation;
import jakarta.validation.Validator;
import jakarta.validation.ValidatorFactory;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public abstract class ProfileServiceImpl<E extends ProfileRepo<U>, U extends Profile> implements ProfileService<U> {

    private final Validator validator;
    protected final E profileRepo;
    protected final ApplicationEventPublisher applicationEventPublisher;

    public ProfileServiceImpl(E userProfileRepo, ApplicationEventPublisher applicationEventPublisher) {
        this.profileRepo = userProfileRepo;
        this.applicationEventPublisher = applicationEventPublisher;
        ValidatorFactory factory = Validation.buildDefaultValidatorFactory();
        validator = factory.getValidator();
    }

    @Override
    public U create(U profile) throws Exception {
        validate(profile);
        profile.setId(UUID.randomUUID().toString());
        U saved = profileRepo.save(profile);
        try {
            applicationEventPublisher.publishEvent(new ProfileCreatedEvent(this, saved));
        } catch (Exception e) { /* don't fail create on publish errors */ }
        return saved;
    }

    @Override
    public U update(String profileId, U profile) throws Exception {

        U persistedProfile = profileRepo.findById(profileId)
                .orElseThrow(() -> new Exception("Profile not found"));
        BeanUtils.copyProperties(profile, persistedProfile);

        U saved = profileRepo.save(persistedProfile);
        try {
            applicationEventPublisher.publishEvent(new ProfileUpdatedEvent(this, saved));
        } catch (Exception e) { /* swallow */ }
        return saved;
    }

    @Override
    public void delete(String id) {
        U p = profileRepo.findById(id).orElse(null);
        profileRepo.deleteById(id);
        if (p != null) {
            try {
                applicationEventPublisher.publishEvent(new ProfileDeletedEvent(this, p));
            } catch (Exception e) { /* ignore */ }
        }
    }

    @Override
    public U find(String profileId) {
        return profileRepo.findById(profileId).orElse(null);
    }

    @Override
    public List<U> findAll() {
        return profileRepo.findAll();
    }

    protected void validate(Object profile) throws Exception {
        Set<ConstraintViolation<Object>> violations = validator.validate(profile);
        if(violations.size() > 0) {
            throw new Exception(violations.iterator().next().getMessage());
        }
    }

    public List<U> findByRole(ProfileRoles role) {
        return profileRepo.findByRole(role);
    }
}

File: izinga-usermanagement/src/main/kotlin/io/curiousoft/izinga/usermanagement/users/ProfileServiceImpl.kt
package io.curiousoft.izinga.usermanagement.users

import io.curiousoft.izinga.commons.model.Profile
import io.curiousoft.izinga.commons.model.ProfileRoles
import io.curiousoft.izinga.commons.profile.events.ProfileCreatedEvent
import io.curiousoft.izinga.commons.profile.events.ProfileDeletedEvent
import io.curiousoft.izinga.commons.profile.events.ProfileUpdatedEvent
import io.curiousoft.izinga.commons.repo.ProfileRepo
import io.curiousoft.izinga.commons.repo.UserProfileRepo
import jakarta.validation.Validation
import jakarta.validation.Validator
import org.springframework.beans.BeanUtils
import org.springframework.context.ApplicationEventPublisher
import java.util.UUID

abstract class ProfileServiceImpl<E : ProfileRepo<U>, U : Profile>(protected val profileRepo: E,
                                                                   private val eventPublisher: ApplicationEventPublisher) : ProfileService<U> {

    private val validator: Validator

    init {
        val factory = Validation.buildDefaultValidatorFactory()
        validator = factory.validator
    }

    @Throws(Exception::class)
    override fun create(profile: U): U {
        validate(profile)
        profile.id = UUID.randomUUID().toString()
        val saved = profileRepo.save<U>(profile)
        eventPublisher.publishEvent(ProfileCreatedEvent(this, saved))
        return saved
    }

    @Throws(Exception::class)
    override fun update(profileId: String, profile: U): U {
        val persistedProfile = profileRepo.findById(profileId)
            .orElseThrow { Exception("Profile not found") }
        BeanUtils.copyProperties(profile, persistedProfile)
        val saved = profileRepo.save(persistedProfile)
        eventPublisher.publishEvent(ProfileUpdatedEvent(this, saved))
        return saved
    }

    override fun delete(id: String) {
        val p = profileRepo.findById(id).orElse(null)
        profileRepo.deleteById(id)
        p?.let { eventPublisher.publishEvent(ProfileDeletedEvent(this, it)) }
    }

    override fun find(profileId: String): U? {
        return profileRepo.findById(profileId).orElse(null)
    }

    override fun findAll(): List<U> {
        return profileRepo.findAll()
    }

    @Throws(Exception::class)
    protected fun validate(profile: Any) {
        val violations = validator.validate(profile)
        if (violations.size > 0) {
            throw Exception(violations.iterator().next().message)
        }
    }

    fun findByRole(role: ProfileRoles?): List<U> {
        return profileRepo.findByRole(role!!)!!
    }
}
